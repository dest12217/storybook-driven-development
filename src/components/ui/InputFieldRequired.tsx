import * as React from 'react';
import ElementInput from '@/components/element/Input';
import ElementInputInvalid from '@/components/element/InputInvalid';
import ElementTextValidationState from '@/components/element/TextValidationState';
import TextLabel from '@/components/element/TextLabel';

interface Props extends React.ComponentProps<'input'> {
  invalid: boolean;
  label: string;
}

type ValidationStateProps = {
  view: boolean;
};

function TextValidationState({ view = false }: ValidationStateProps) {
  if (!view) {
    return null;
  }

  return <ElementTextValidationState>This field is required</ElementTextValidationState>;
}

interface InputProps extends React.ComponentProps<'input'> {
  invalid: boolean;
}

function Input({ invalid, ...props }: InputProps) {
  if (invalid) {
    return <ElementInputInvalid {...props} />;
  }

  return <ElementInput {...props} />;
}

export default function InputFieldRequired({ label, invalid, ...props }: Props) {
  return <>
    <TextLabel>{label} [Required]</TextLabel>
    <Input invalid={invalid} {...props} />
    <TextValidationState view={invalid} />
  </>
}
