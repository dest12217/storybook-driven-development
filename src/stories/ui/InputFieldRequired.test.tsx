/// <reference types="@testing-library/jest-dom" />
import { composeStories } from '@storybook/react';
import { render } from '@testing-library/react';
import type { ComposedStoryFn } from 'storybook/internal/types';
import { describe, expect, test } from 'vitest';
import * as stories from '@/stories/ui/InputFieldRequired.stories';

const composedStories = composeStories(stories);

describe('InputFieldRequired', () => {
  describe('Snapshot', () => {
    test.each(Object.entries<ComposedStoryFn>(composedStories))('Snapshot: %s', (name, Story) => {
      const { asFragment } = render(<Story />);

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('Invalid', () => {
    test('エラーメッセージが表示されていること', () => {
      const { Invalid } = composedStories;
      const { getAllByText } = render(<Invalid />);
      const elements = getAllByText('This field is required');

      expect(elements.length).toBeGreaterThan(0);
      elements.forEach((element) => {
        expect(element).toBeVisible();
      });
    });
  });
});
