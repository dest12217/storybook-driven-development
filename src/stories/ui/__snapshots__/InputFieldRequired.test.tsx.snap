// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InputFieldRequired > Snapshot > Snapshot: Invalid 1`] = `
<DocumentFragment>
  <label
    class="element-text-label"
  >
    label [Required]
  </label>
  <input
    class="element-input-invalid"
  />
  <p
    aria-live="assertive"
    class="element-text-validation-state"
  >
    This field is required
  </p>
</DocumentFragment>
`;

exports[`InputFieldRequired > Snapshot > Snapshot: Valid 1`] = `
<DocumentFragment>
  <label
    class="element-text-label"
  >
    label [Required]
  </label>
  <input
    class="element-input"
  />
</DocumentFragment>
`;
